client/auth/login.jsx

```diff
@@ -103,27 +101,58 @@ module.exports = React.createClass( {
 	},
 
 	canSubmitForm: function() {
+		const auth = this.state.auth;
+
 		// No submission until the ajax has finished
-		if ( this.state.inProgress ) {
+		if ( auth.inProgress ) {
 			return false;
 		}
 
 		// If we have 2fa set then don't allow submission until a code is entered
-		if ( this.state.requires2fa ) {
+		if ( auth.required2faType ) {
 			return parseInt( this.state.auth_code, 10 ) > 0;
 		}
 
 		// Don't allow submission until username+password is entered
 		return this.hasLoginDetails();
 	},
 
-	toggleSelfHostedInstructions: function () {
-	    var isShowing = !this.state.showInstructions;
-	    this.setState( { showInstructions: isShowing } );
+	toggleSelfHostedInstructions: function() {
+		const isShowing = ! this.state.showInstructions;
+		this.setState( { showInstructions: isShowing } );
+	},
+
+	useAuthCode: function() {
+		AuthActions.useAuthCode(
+			this.translate( 'Please enter the verification code generated by your Authenticator mobile application.' )
+		);
+	},
+
+	focusWhenIdle: function( ref ) {
+		if ( this.state.auth.inProgress === false ) {
+			ref.focus();
+		}
+	},
+
+	updateLogin: function( e ) {
+		this.setState( { login: e.target.value } );
+	},
+
+	updatePassword: function( e ) {
+		this.setState( { password: e.target.value } );
+	},
+
+	updateAuthCode: function( e ) {
+		this.setState( { auth_code: e.target.value } );
 	},
```

Comment:
With these three methods I wonder if we couldn't combine them into a more generic updater…

```is
storeEventValueIn( stateKey ) {
	return function( { target: value } ) {
		this.setState( { [ stateKey ]: value } );
	}
}

<FormTextInput onChange={ this.storeEventValueIn( 'login' ) } />
<FormTextInput onChange={ this.storeEventValueIn( 'password' ) } />
<FormTextInput onChange={ this.storeEventValueIn( 'auth_code' ) } />
```